neqo-bin/src/server/mod.rs

```diff
@@ -198,240 +201,146 @@ trait HttpServer: Display {
-    offset: usize,
-    remaining: usize,
-}
-
-impl From<&[u8]> for ResponseData {
-    fn from(data: &[u8]) -> Self {
-        Self::from(data.to_vec())
-    }
-}
-
-impl From<Vec<u8>> for ResponseData {
-    fn from(data: Vec<u8>) -> Self {
-        let remaining = data.len();
-        Self {
-            data: Cow::Owned(data),
-            offset: 0,
-            remaining,
-        }
-    }
-}
-
-impl ResponseData {
-    fn repeat(buf: &'static [u8], total: usize) -> Self {
-        Self {
-            data: Cow::Borrowed(buf),
-            offset: 0,
-            remaining: total,
-        }
-    }
-
-    fn send(&mut self, stream: &mut Http3OrWebTransportStream) {
-        while self.remaining > 0 {
-            let end = min(self.data.len(), self.offset + self.remaining);
-            let slice = &self.data[self.offset..end];
-            match stream.send_data(slice) {
-                Ok(0) => {
-                    return;
-                }
-                Ok(sent) => {
-                    self.remaining -= sent;
-                    self.offset = (self.offset + sent) % self.data.len();
-                }
-                Err(e) => {
-                    qwarn!("Error writing to stream {}: {:?}", stream, e);
-                    return;
-                }
-            }
-        }
-    }
-
-    fn done(&self) -> bool {
-        self.remaining == 0
+    fn get_timeout(&self) -> Option<Duration> {
+        None
     }
 }
 
-struct SimpleServer {
-    server: Http3Server,
-    /// Progress writing to each stream.
-    remaining_data: HashMap<StreamId, ResponseData>,
-    posts: HashMap<Http3OrWebTransportStream, usize>,
+enum ServerType {
+    Http3,
+    Http3Fail,
+    Http3NoResponse,
+    Http3Ech,
+    Http3Proxy,
 }
 
-impl SimpleServer {
-    const MESSAGE: &'static [u8] = &[0; 4096];
-
-    pub fn new(
-        args: &Args,
-        anti_replay: AntiReplay,
-        cid_mgr: Rc<RefCell<dyn ConnectionIdGenerator>>,
-    ) -> Self {
-        let server = Http3Server::new(
-            args.now(),
-            &[args.key.clone()],
-            &[args.shared.alpn.clone()],
-            anti_replay,
-            cid_mgr,
-            Http3Parameters::default()
-                .connection_parameters(args.shared.quic_parameters.get(&args.shared.alpn))
-                .max_table_size_encoder(args.shared.max_table_size_encoder)
-                .max_table_size_decoder(args.shared.max_table_size_decoder)
-                .max_blocked_streams(args.shared.max_blocked_streams),
-            None,
-        )
-        .expect("We cannot make a server!");
-        Self {
-            server,
-            remaining_data: HashMap::new(),
-            posts: HashMap::new(),
-        }
-    }
-}
-
-impl Display for SimpleServer {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.server.fmt(f)
-    }
+// TODO: Use singular form.
+struct ServersRunner {
+    args: Args,
+    server: Box<dyn HttpServer>,
```

Comment:
Note that we could have multiple servers here.
https://searchfox.org/mozilla-central/rev/e69f323af80c357d287fb6314745e75c62eab92a/netwerk/test/http3server/src/main.rs#1090

Reply:
Instead of one `ServersRunner` to run multiple `HttpServer`s, I simply instantiate multiple `ServersRunners` in the mozilla-central `http3server`.

(And I renamed `ServersRunner` to `ServerRunner`.)

See `main` `fn` at the very bottom of the diff:

https://phabricator.services.mozilla.com/D209574